# Listed Data EDA (Week2)
Using the dataset filtered "PropertyType == Residential"
Author : Yoora Choi

In [14]:
import pandas as pd
import numpy as np
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)

In [15]:
# Import listed data (2024.01-2026.03)
listed = pd.read_csv("/Users/yoorachoi/Python/IDX/Listing/data/CRMLSListing_merged.csv",  low_memory=False)

## 1.Data overview

In [16]:
# Check data structure
listed.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 540612 entries, 0 to 540611
Data columns (total 84 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   OriginalListPrice             539845 non-null  float64
 1   ListingKey                    540612 non-null  int64  
 2   ListAgentEmail                493182 non-null  object 
 3   CloseDate                     161735 non-null  object 
 4   ClosePrice                    141190 non-null  float64
 5   ListAgentFirstName            536362 non-null  object 
 6   ListAgentLastName             540572 non-null  object 
 7   Latitude                      460444 non-null  float64
 8   Longitude                     460444 non-null  float64
 9   UnparsedAddress               539919 non-null  object 
 10  PropertyType                  540612 non-null  object 
 11  LivingArea                    540057 non-null  float64
 12  ListPrice                     540612 non-nul

In [17]:
# check the data structure
listed.head(5)

,OriginalListPrice,ListingKey,ListAgentEmail,CloseDate,ClosePrice,ListAgentFirstName,ListAgentLastName,Latitude,Longitude,UnparsedAddress,PropertyType,LivingArea,ListPrice,DaysOnMarket,ListOfficeName,BuyerOfficeName,CoListOfficeName,ListAgentFullName,CoListAgentFirstName,CoListAgentLastName,BuyerAgentMlsId,BuyerAgentFirstName,BuyerAgentLastName,FireplacesTotal,AssociationFeeFrequency,AboveGradeFinishedArea,ListingKeyNumeric,MLSAreaMajor,TaxAnnualAmount,CountyOrParish,PropertyType.1,MlsStatus,ElementarySchool,ListAgentFirstName.1,AttachedGarageYN,ParkingTotal,BuilderName,PropertySubType,LotSizeAcres,SubdivisionName,BuyerOfficeAOR,YearBuilt,DaysOnMarket.1,BuyerAgencyCompensationType,StreetNumberNumeric,LivingArea.1,ListingId,BathroomsTotalInteger,City,BuyerAgencyCompensation,TaxYear,BuildingAreaTotal,BedroomsTotal,ContractStatusChangeDate,Longitude.1,ElementarySchoolDistrict,CoBuyerAgentFirstName,PurchaseContractDate,ListingContractDate,BelowGradeFinishedArea,BusinessType,Latitude.1,ListPrice.1,StateOrProvince,CoveredSpaces,MiddleOrJuniorSchool,FireplaceYN,Stories,HighSchool,Levels,ListAgentLastName.1,CloseDate.1,LotSizeDimensions,LotSizeArea,MainLevelBedrooms,NewConstructionYN,GarageSpaces,HighSchoolDistrict,PostalCode,BuyerOfficeName.1,AssociationFee,LotSizeSquareFeet,MiddleOrJuniorSchoolDistrict,UnparsedAddress.1
0,1340000.0,1074973329,haleh360@Gmail.com,NaN,NaN,Haleh,Dowlatshahi,34.052207,-118.408445,2220 Avenue Of The Stars 2704,Residential,1301.0,1340000.0,127,Rodeo Realty- Brentwood,NaN,NaN,Haleh Dowlatshahi,NaN,NaN,NaN,NaN,NaN,NaN,Monthly,NaN,1074973329,C05 - Westwood - Century City,NaN,Los Angeles,Residential,Pending,NaN,Haleh,NaN,0.0,NaN,Condominium,4.0831,NaN,NaN,1964.0,127,Item1,2220.0,1301.0,24389291,2.0,Los Angeles,2.0,NaN,1301.0,2.0,2024-05-07,-118.408445,NaN,NaN,2024-05-07,2024-01-01,NaN,NaN,34.052207,1340000.0,CA,NaN,NaN,False,1.0,NaN,One,Dowlatshahi,NaN,NaN,177861.0,NaN,False,NaN,NaN,90067,NaN,2105.00,177861.0,NaN,2220 Avenue Of The Stars 2704
1,2500000.0,1074954552,Reneechen@yourhomesoldguaranteed.com,NaN,NaN,Renee,Chen,33.496363,-117.691677,16 Palisades,Residential,2788.0,2500000.0,1,Your Home Sold Guaranteed Realty,NaN,NaN,Renee Chen,NaN,NaN,NaN,NaN,NaN,NaN,Monthly,NaN,1074954552,LNSLT - Salt Creek,NaN,Orange,Residential,Active,NaN,Renee,True,3.0,NaN,SingleFamilyResidence,0.1217,Warmington Homes (BB) (BBW),NaN,1994.0,1,Item1,16.0,2788.0,AR24091628,3.0,Laguna Niguel,2.0,NaN,NaN,4.0,2024-01-24,-117.691677,NaN,NaN,NaN,2024-01-24,NaN,NaN,33.496363,2500000.0,CA,NaN,NaN,True,2.0,NaN,Two,Chen,NaN,NaN,5300.0,0.0,False,3.0,Capistrano Unified,92677,NaN,254.00,5300.0,NaN,16 Palisades
2,3150000.0,1074936537,anader@dppre.com,NaN,NaN,Margaret,Nader,34.119345,-118.111254,1615 Waverly Road,Residential,3250.0,3150000.0,1,COMPASS,NaN,NaN,Margaret Nader,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1074936537,655 - San Marino,NaN,Los Angeles,Residential,Active,NaN,Margaret,True,2.0,NaN,SingleFamilyResidence,0.2159,NaN,NaN,1965.0,1,Item1,1615.0,3250.0,P1-17547,4.0,San Marino,2.5,NaN,NaN,6.0,2024-01-12,-118.111254,NaN,NaN,NaN,2024-01-12,NaN,NaN,34.119345,3150000.0,CA,NaN,NaN,True,2.0,NaN,Two,Nader,NaN,NaN,9404.0,NaN,NaN,2.0,NaN,91108,NaN,NaN,9404.0,NaN,1615 Waverly Road
3,3090000.0,1074917818,QIANYU0607@GMAIL.COM,NaN,NaN,QIANYU,GUAN,33.984057,-117.802819,2250 Indian Creek Road,Residential,7456.0,3090000.0,0,"Signature One Realty Group, Inc",NaN,NaN,QIANYU GUAN,NaN,NaN,NaN,NaN,NaN,NaN,Monthly,NaN,1074917818,616 - Diamond Bar,NaN,Los Angeles,Residential,Active,Castle Rock,QIANYU,True,4.0,NaN,SingleFamilyResidence,1.3368,NaN,NaN,2003.0,0,Item1,2250.0,7456.0,OC24091496,9.0,Diamond Bar,2.0,NaN,NaN,6.0,2024-01-20,-117.802819,NaN,NaN,NaN,2024-01-20,NaN,NaN,33.984057,3090000.0,CA,NaN,Chaparral,True,NaN,Diamond Bar,ThreeOrMore,GUAN,NaN,NaN,58232.0,1.0,False,4.0,Walnut Valley Unified,91765,NaN,295.95,58232.0,NaN,2250 Indian Creek Road
4,12725000.0,1074143166,jeff.williams@pacificsir.com,NaN,NaN,Jeff,Williams,33.607583,-117.887743,317 E. Bayfront,Reside

In [18]:
# Summary statistics for numerical variable
listed.describe()

,OriginalListPrice,ListingKey,ClosePrice,Latitude,Longitude,LivingArea,ListPrice,DaysOnMarket,FireplacesTotal,AboveGradeFinishedArea,ListingKeyNumeric,TaxAnnualAmount,ParkingTotal,LotSizeAcres,YearBuilt,DaysOnMarket.1,StreetNumberNumeric,LivingArea.1,BathroomsTotalInteger,BuyerAgencyCompensation,TaxYear,BuildingAreaTotal,BedroomsTotal,Longitude.1,ElementarySchoolDistrict,BelowGradeFinishedArea,BusinessType,Latitude.1,ListPrice.1,CoveredSpaces,Stories,LotSizeArea,MainLevelBedrooms,GarageSpaces,AssociationFee,LotSizeSquareFeet,MiddleOrJuniorSchoolDistrict
count,5.398450e+05,5.406120e+05,1.411900e+05,460444.000000,460444.000000,5.400570e+05,5.406120e+05,540612.000000,0.0,0.0,5.406120e+05,0.0,5.405920e+05,4.960030e+05,539679.000000,540612.000000,5.397390e+05,5.400570e+05,540557.000000,78993.000000,0.0,48111.000000,540465.000000,460444.000000,0.0,3023.000000,0.0,460444.000000,5.406120e+05,0.0,445876.000000,4.968410e+05,293647.000000,510527.000000,410446.000000,4.966670e+05,0.0
mean,1.398189e+06,1.104832e+09,1.201480e+06,34.606384,-118.463944,1.979979e+03,1.314312e+06,19.139886,NaN,NaN,1.104832e+09,NaN,7.618042e+00,6.518526e+01,1979.598326,19.139886,1.158825e+04,1.979979e+03,2.626204,189.109913,NaN,2445.258652,3.222923,-118.463944,NaN,87.238574,NaN,34.606384,1.314312e+06,NaN,1.370991,4.692671e+04,2.046879,1.844685,271.152047,4.103986e+05,NaN
std,7.354788e+06,2.989437e+07,4.354100e+06,2.000012,3.555755,2.337340e+04,2.348640e+06,26.682084,NaN,NaN,2.989437e+07,NaN,3.527765e+03,1.213267e+04,27.000918,26.682084,2.853986e+05,2.337340e+04,3.256854,2340.716879,NaN,2248.077581,1.186857,3.555755,NaN,360.500960,NaN,2.000012,2.348640e+06,NaN,0.483071,2.178064e+06,6.442479,3.877160,2369.383731,2.079520e+07,NaN
min,0.000000e+00,1.018357e+09,5.250000e+02,-117.472493,-159.475987,0.000000e+00,1.000000e+02,-58.000000,NaN,NaN,1.018357e+09,NaN,-1.430000e+02,0.000000e+00,1776.000000,-58.000000,0.000000e+00,0.000000e+00,0.000000,0.000000,NaN,0.000000,0.000000,-159.475987,NaN,0.000000,NaN,-117.472493,1.000000e+02,NaN,1.000000,0.000000e+00,0.000000,0.000000,0.000000,0.000000e+00,NaN
25%,5.850000e+05,1.077381e+09,6.000000e+05,33.740149,-118.696760,1.247000e+03,5.800000e+05,5.000000,NaN,NaN,1.077381e+09,NaN,2.000000e+00,1.200000e-01,1961.000000,5.000000,9.750000e+02,1.247000e+03,2.000000,2.000000,NaN,1305.000000,2.000000,-118.696760,NaN,0.000000,NaN,33.740149,5.800000e+05,NaN,1.000000,5.000000e+03,1.000000,1.000000,0.000000,5.227000e+03,NaN
50%,8.490000e+05,1.104932e+09,8.500000e+05,34.053944,-118.029501,1.669000e+03,8.450000e+05,11.000000,NaN,NaN,1.104932e+09,NaN,2.000000e+00,1.700000e-01,1980.000000,11.000000,3.266000e+03,1.669000e+03,2.000000,2.500000,NaN,1825.000000,3.000000,-118.029501,NaN,0.000000,NaN,34.053944,8.450000e+05,NaN,1.000000,7.194000e+03,2.000000,2.000000,125.000000,7.405000e+03,NaN
75%,1.389999e+06,1.126964e+09,1.350000e+06,34.478232,-117.237676,2.300000e+03,1.375000e+06,22.000000,NaN,NaN,1.126964e+09,NaN,3.000000e+00,3.137000e-01,2001.000000,22.000000,1.204600e+04,2.300000e+03,3.000000,2.500000,NaN,2800.000000,4.000000,-117.237676,NaN,0.000000,NaN,34.478232,1.375000e+06,NaN,2.000000,1.204800e+04,3.000000,2.000000,400.000000,1.360300e+04,NaN
max,1.390000e+09,1.156489e+09,8.200000e+08,737.000000,329.000000,1.702132e+07,1.950000e+08,731.000000,NaN,NaN,1.156489e+09,NaN,2.593669e+06,4.187292e+06,2028.000000,731.000000,7.954796e+07,1.702132e+07,2208.000000,200000.000000,NaN,123764.000000,94.000000,329.000000,NaN,5000.000000,NaN,737.000000,1.950000e+08,NaN,2.000000,9.187423e+08,3333.000000,600.000000,968348.000000,7.638115e+09,NaN


In [19]:
# Summary statistics for non-numerical variable
listed.describe(include = 'object')

,ListAgentEmail,CloseDate,ListAgentFirstName,ListAgentLastName,UnparsedAddress,PropertyType,ListOfficeName,BuyerOfficeName,CoListOfficeName,ListAgentFullName,CoListAgentFirstName,CoListAgentLastName,BuyerAgentMlsId,BuyerAgentFirstName,BuyerAgentLastName,AssociationFeeFrequency,MLSAreaMajor,CountyOrParish,PropertyType.1,MlsStatus,ElementarySchool,ListAgentFirstName.1,AttachedGarageYN,BuilderName,PropertySubType,SubdivisionName,BuyerOfficeAOR,BuyerAgencyCompensationType,ListingId,City,ContractStatusChangeDate,CoBuyerAgentFirstName,PurchaseContractDate,ListingContractDate,StateOrProvince,MiddleOrJuniorSchool,FireplaceYN,HighSchool,Levels,ListAgentLastName.1,CloseDate.1,LotSizeDimensions,NewConstructionYN,HighSchoolDistrict,PostalCode,BuyerOfficeName.1,UnparsedAddress.1
count,493182,161735,536362,540572,539919,540612,540612,152758,121031,540521,120546,120946,154840,154237,154947,231037,467425,540612,540612,540612,64355,536362,446096,25233,539356,199023,142941,79013,540612,540037,534503,14001,263213,540612,540559,64485,540038,84456,482090,540572,161735,28227,503509,377873,540599,152758,539919
unique,89565,906,17035,37396,474831,1,20688,13637,6147,86171,7447,15339,70459,13649,27863,4,1108,63,1,5,1424,17035,2,3797,21,17297,66,3,540517,1208,844,2755,826,821,29,672,2,529,18,37396,906,11963,2,451,4356,13637,474831
top,randy.anderson@richmondamericanhomes.com,2024-03-29,Michael,Lee,5166 Via Portola,Residential,Compass,Compass,Compass,Todd Myatt,Michael,Myatt,NONMEMBER,General,NONMEMBER,Monthly,699 - Not Defined,Los Angeles,Residential,Active,Other,Michael,True,Lennar,SingleFamilyResidence,Other,SanDiego,Item1,NDP2501178,Los Angeles,2024-04-30,Michael,2024-04-24,2025-05-01,CA,Other,True,Other,One,Lee,2024-03-29,50x135,False,Other,92253,Compass,5166 Via Portola
freq,1091,1057,6788,2636,12,540612,38528,12158,14085,1133,1775,1134,2409,1980,2012,216924,61554,137333,540612,269890,13248,6788,349988,2356,394249,5430,11187,77938,4,27925,2372,191,880,1776,540404,9878,348553,4107,277502,2636,1057,482,481718,44906,4006,12158,12


## 2. Missing data check(null)

In [20]:
# null count
null_count = listed.isnull().sum()

# null percentage (%)
null_percent = (listed.isnull().sum() / len(listed)) * 100

# data frame
null_summary = pd.DataFrame({
    'null_count': null_count,
    'null_percent': null_percent
}).sort_values(by='null_percent', ascending=False)

null_summary['null_percent'] = null_summary['null_percent'].round(2)
null_summary

,null_count,null_percent
TaxAnnualAmount,540612,100.00
FireplacesTotal,540612,100.00
ElementarySchoolDistrict,540612,100.00
TaxYear,540612,100.00
BusinessType,540612,100.00
MiddleOrJuniorSchoolDistrict,540612,100.00
CoveredSpaces,540612,100.00
AboveGradeFinishedArea,540612,100.00
BelowGradeFinishedArea,537589,99.44
CoBuyerAgentFirstName,526611,97.41


## 3. Check duplicates

### 3.1 duplicated columns

In [21]:
# check duplicate columns
pairs = []

for col in listed.columns:
    if '.' in col:
        base_col = col.split('.')[0]
        if base_col in listed.columns:
            pairs.append((base_col, col))

pairs

[('PropertyType', 'PropertyType.1'),
 ('ListAgentFirstName', 'ListAgentFirstName.1'),
 ('DaysOnMarket', 'DaysOnMarket.1'),
 ('LivingArea', 'LivingArea.1'),
 ('Longitude', 'Longitude.1'),
 ('Latitude', 'Latitude.1'),
 ('ListPrice', 'ListPrice.1'),
 ('ListAgentLastName', 'ListAgentLastName.1'),
 ('CloseDate', 'CloseDate.1'),
 ('BuyerOfficeName', 'BuyerOfficeName.1'),
 ('UnparsedAddress', 'UnparsedAddress.1')]

In [22]:
# Check if duplicate column pairs have identical values row-by-row
for base, dup in pairs:
    same_ratio = (listed[base] == listed[dup]).mean()
    print(f"{base} vs {dup}: {same_ratio:.2f}")

PropertyType vs PropertyType.1: 1.00
ListAgentFirstName vs ListAgentFirstName.1: 0.99
DaysOnMarket vs DaysOnMarket.1: 1.00
LivingArea vs LivingArea.1: 1.00
Longitude vs Longitude.1: 0.85
Latitude vs Latitude.1: 0.85
ListPrice vs ListPrice.1: 1.00
ListAgentLastName vs ListAgentLastName.1: 1.00
CloseDate vs CloseDate.1: 0.30
BuyerOfficeName vs BuyerOfficeName.1: 0.28
UnparsedAddress vs UnparsedAddress.1: 1.00


### 3.2 duplicated rows

In [23]:
# check completely duplicated rows
listed[listed.duplicated(keep=False)]

,OriginalListPrice,ListingKey,ListAgentEmail,CloseDate,ClosePrice,ListAgentFirstName,ListAgentLastName,Latitude,Longitude,UnparsedAddress,PropertyType,LivingArea,ListPrice,DaysOnMarket,ListOfficeName,BuyerOfficeName,CoListOfficeName,ListAgentFullName,CoListAgentFirstName,CoListAgentLastName,BuyerAgentMlsId,BuyerAgentFirstName,BuyerAgentLastName,FireplacesTotal,AssociationFeeFrequency,AboveGradeFinishedArea,ListingKeyNumeric,MLSAreaMajor,TaxAnnualAmount,CountyOrParish,PropertyType.1,MlsStatus,ElementarySchool,ListAgentFirstName.1,AttachedGarageYN,ParkingTotal,BuilderName,PropertySubType,LotSizeAcres,SubdivisionName,BuyerOfficeAOR,YearBuilt,DaysOnMarket.1,BuyerAgencyCompensationType,StreetNumberNumeric,LivingArea.1,ListingId,BathroomsTotalInteger,City,BuyerAgencyCompensation,TaxYear,BuildingAreaTotal,BedroomsTotal,ContractStatusChangeDate,Longitude.1,ElementarySchoolDistrict,CoBuyerAgentFirstName,PurchaseContractDate,ListingContractDate,BelowGradeFinishedArea,BusinessType,Latitude.1,ListPrice.1,StateOrProvince,CoveredSpaces,MiddleOrJuniorSchool,FireplaceYN,Stories,HighSchool,Levels,ListAgentLastName.1,CloseDate.1,LotSizeDimensions,LotSizeArea,MainLevelBedrooms,NewConstructionYN,GarageSpaces,HighSchoolDistrict,PostalCode,BuyerOfficeName.1,AssociationFee,LotSizeSquareFeet,MiddleOrJuniorSchoolDistrict,UnparsedAddress.1


In [24]:
listed.duplicated().sum()

np.int64(0)

### 6. Key MLS Fields – Statistical Summary

In [25]:
# Numerical key fields summary, 50% = median
key_num_cols = ['OriginalListPrice','ListPrice', 'ClosePrice', 'LotSizeAcres', 'LivingArea', 'BedroomsTotal', 'BathroomsTotalInteger', 'DaysOnMarket', 'YearBuilt']
listed[key_num_cols].describe().round(2)

,OriginalListPrice,ListPrice,ClosePrice,LotSizeAcres,LivingArea,BedroomsTotal,BathroomsTotalInteger,DaysOnMarket,YearBuilt
count,5.398450e+05,5.406120e+05,1.411900e+05,496003.00,540057.00,540465.00,540557.00,540612.00,539679.0
mean,1.398189e+06,1.314312e+06,1.201480e+06,65.19,1979.98,3.22,2.63,19.14,1979.6
std,7.354788e+06,2.348640e+06,4.354100e+06,12132.67,23373.40,1.19,3.26,26.68,27.0
min,0.000000e+00,1.000000e+02,5.250000e+02,0.00,0.00,0.00,0.00,-58.00,1776.0
25%,5.850000e+05,5.800000e+05,6.000000e+05,0.12,1247.00,2.00,2.00,5.00,1961.0
50%,8.490000e+05,8.450000e+05,8.500000e+05,0.17,1669.00,3.00,2.00,11.00,1980.0
75%,1.389999e+06,1.375000e+06,1.350000e+06,0.31,2300.00,4.00,3.00,22.00,2001.0
max,1.390000e+09,1.950000e+08,8.200000e+08,4187292.00,17021321.00,94.00,2208.00,731.00,2028.0


# Listed Data EDA — Week 2 Summary
**Dataset:** CRMLS Residential Listings (Jan 2024 – Mar 2026)  
**Filter Applied:** `PropertyType == Residential`  
**Author:** Yoora Choi

---

## 1. Dataset Overview

| Item | Detail |
|------|--------|
| Total Rows | 540,611 |
| Total Columns | 84 |
| Data Period | January 2024 – March 2026 |
| Data Types | float64: 32 / int64: 4 / object: 48 |

---

## 2. Missing Value Analysis

###   100% Missing (8 columns) (Drop Immediately)

| Column | Missing Rate |
|--------|-------------|
| `CoveredSpaces` | 100% |
| `MiddleOrJuniorSchoolDistrict` | 100% |
| `ElementarySchoolDistrict` | 100% |
| `AboveGradeFinishedArea` | 100% |
| `FireplacesTotal` | 100% |
| TaxAnnualAmount` | 100% |
| TaxYear` | 100% |
| BusinessType` | 100% |

---

###  90–100% Missing (5 columns)

| Column | Missing Rate |
|--------|-------------|
| `BelowGradeFinishedArea` | 99.44% |
| `CoBuyerAgentFirstName` | 97.41% |
| `BuilderName` | 95.33% |
| `LotSizeDimensions` | 94.78% |
| `BuildingAreaTotal` | 91.10% |

---

### 70–80% Missing 

| Column | Missing Rate |
|--------|-------------|
| `CoListAgentFirstName` | 80.05% | 
| `CoListAgentLastName` | 80.00% | 
| `CoListOfficeName` | 79.98% | 
| `ClosePrice` | 72.74% | 
| `BuyerOfficeAOR` | 72.66% | 
| `BuyerOfficeName` | 71.28% | 
| `AssociationFeeFrequency` | 71.14% | 
| `BuyerAgentFirstName` | 71.11% | 
| `BuyerAgentMlsId` | 71.09% | 
| `BuyerAgentLastName` | 71.01% |
| `CloseDate` | 70.06% |

> **These are expected nulls.** Buyer-related and close-related fields are only populated when a transaction has closed. Missing values here reflect Active/Pending listings, not data quality issues.

---

### 10–70% Missing

| Column | Missing Rate |
|--------|-------------|
| `SubdivisionName` | 66.61% |
| `PurchaseContractDate` | 56.89% |
| `MainLevelBedrooms` | 53.76% |
| `HighSchoolDistrict` | 39.76% |
| `AttachedGarageYN` | 34.09% |
| `AssociationFee` | 32.32% |
| `Stories` | 25.94% |
| `Levels` | 19.22% |
| `GarageSpaces` | 18.99% |
| `Latitude / Longitude` | ~13% |
| `LivingArea` | 12.47% |
| `PropertySubType` | 12.20% |
| `BedroomsTotal` | 12.07% |
| `MLSAreaMajor` | 11.29% |

---

### 0–10% Missing

| Column | Missing Rate |
|--------|-------------|
| `LotSizeAcres` | 9.55% |
| `LotSizeSquareFeet` | 9.24% |
| `LotSizeArea` | 9.11% |
| `ListAgentEmail` | 8.60% |
| `BathroomsTotalInteger` | 8.19% |
| `YearBuilt` | 8.14% |
| `ParkingTotal` | 6.13% |
| `OriginalListPrice` | 0.39% |
| `City` | 0.11% |
| `ListAgentFullName` | 0.03% |
| `StateOrProvince` | 0.01% |

### No Missing Values

`DaysOnMarket`, `PropertyType`, `ListPrice`, `ListingKey`, `ListingKeyNumeric`, `ListOfficeName`, `MlsStatus`, `ListingId`, `ListingContractDate`, `CountyOrParish`

---

## 3. Duplicate Analysis

### Duplicate Rows

> **No fully duplicate rows found.** Row-level uniqueness is confirmed.

---

### Duplicate Column Pairs — 11 Pairs Found

| Match Rate | Column Pairs | Action |
|-----------|-------------|--------|
| 100% | `PropertyType` / `PropertyType.1` | Drop `.1` |
| 100% | `DaysOnMarket` / `DaysOnMarket.1` | Drop `.1` |
| 100% | `ListPrice` / `ListPrice.1` | Drop `.1` |
| 100% | `ListAgentLastName` / `ListAgentLastName.1` | Drop `.1` |
| 100% | `UnparsedAddress` / `UnparsedAddress.1` |  Drop `.1` |
| ~99% | `ListAgentFirstName` / `ListAgentFirstName.1` | Drop `.1` |
| ~87–88% | `Latitude` / `Latitude.1` |  Drop `.1` |
| ~87–88% | `Longitude` / `Longitude.1` | Drop `.1` |
| ~87–88% | `LivingArea` / `LivingArea.1` | Drop `.1` |
| ~29–30% | `CloseDate` / `CloseDate.1` |  Values differ — inspect before dropping |
| ~29–30% | `BuyerOfficeName` / `BuyerOfficeName.1` |  Values differ — inspect before dropping |

>  `CloseDate.1` and `BuyerOfficeName.1` have very low match rates (~29–30%), suggesting they may contain different data snapshots. Do not drop without further investigation.

---

## 4. Key Numeric Field Distributions

In [26]:


key_num_cols = ['OriginalListPrice', 'ListPrice', 'ClosePrice', 'LotSizeAcres',
                'LivingArea', 'BedroomsTotal', 'BathroomsTotalInteger',
                'DaysOnMarket', 'YearBuilt']
listed[key_num_cols].describe().round(2)



,OriginalListPrice,ListPrice,ClosePrice,LotSizeAcres,LivingArea,BedroomsTotal,BathroomsTotalInteger,DaysOnMarket,YearBuilt
count,5.398450e+05,5.406120e+05,1.411900e+05,496003.00,540057.00,540465.00,540557.00,540612.00,539679.0
mean,1.398189e+06,1.314312e+06,1.201480e+06,65.19,1979.98,3.22,2.63,19.14,1979.6
std,7.354788e+06,2.348640e+06,4.354100e+06,12132.67,23373.40,1.19,3.26,26.68,27.0
min,0.000000e+00,1.000000e+02,5.250000e+02,0.00,0.00,0.00,0.00,-58.00,1776.0
25%,5.850000e+05,5.800000e+05,6.000000e+05,0.12,1247.00,2.00,2.00,5.00,1961.0
50%,8.490000e+05,8.450000e+05,8.500000e+05,0.17,1669.00,3.00,2.00,11.00,1980.0
75%,1.389999e+06,1.375000e+06,1.350000e+06,0.31,2300.00,4.00,3.00,22.00,2001.0
max,1.390000e+09,1.950000e+08,8.200000e+08,4187292.00,17021321.00,94.00,2208.00,731.00,2028.0


### Price Fields

| Field | Count | Median | Mean | Min | Max | Flag |
|-------|-------|--------|------|-----|-----|------|
| `OriginalListPrice` | 539,845 | $849,000 | $1.40M | $0 | $1.39B | 🔴 Outliers |
| `ListPrice` | 540,612 | $845,000 | $1.31M | $100 | $195M | 🔴 Outliers |
| `ClosePrice` | 141,190 | $850,000 | $1.20M | $525 | $820M | 🟠 Skewed |

> `ClosePrice` covers only ~26% of all records (closed listings only). The mean is ~$350K above the median — a sign of strong right skew from high-end luxury sales. **Median ($850K) is the more representative market benchmark.**

---

### Property Attribute Fields

| Field | Count | Median | Mean | Min | Max | Flag |
|-------|-------|--------|------|-----|-----|------|
| `LivingArea` | 540,057 | 1,669 sqft | 1,980 sqft | 0 | 17,021,321 sqft | 🔴 Outliers |
| `LotSizeAcres` | 496,003 | 0.17 ac | 65.19 ac | 0 | 4,187,292 ac | 🔴 Outliers |
| `BedroomsTotal` | 540,465 | 3 | 3.22 | 0 | 94 | 🟠 Review |
| `BathroomsTotalInteger` | 540,557 | 2 | 2.63 | 0 | 2,208 | 🔴 Outliers |
| `DaysOnMarket` | 540,612 | 11 | 19.14 | −58 | 731 | 🔴 Outliers |
| `YearBuilt` | 539,679 | 1980 | 1979.6 | 1776 | 2028 | 🟠 Review |

**Outlier explanations:**

| Value | Likely Cause |
|-------|-------------|
| `OriginalListPrice` = $0 | Missing or placeholder — remove |
| `LivingArea` = 17M sqft | Unit mismatch or data entry error |
| `LotSizeAcres` = 4.19M ac | Unit error or erroneous entry |
| `DaysOnMarket` = −58 | Calculation error — must be removed |
| `DaysOnMarket` = 731 | 2 full years on market — verify or cap |
| `BathroomsTotalInteger` = 2,208 | Clear data entry error |
| `BedroomsTotal` = 94 | Possible commercial misclassification |
| `YearBuilt` = 2028 | Future-dated new construction entry |
| `YearBuilt` = 1776 | Likely a placeholder value |

---

## 5. Intern Questions — Answered

### Q1. What is the Residential vs. other property type share?



In [27]:
# PropertySubType distribution
subtype_dist = listed['PropertySubType'].value_counts(dropna=False)
subtype_pct = listed['PropertySubType'].value_counts(normalize=True, dropna=False) * 100

subtype_summary = pd.DataFrame({
    'count': subtype_dist,
    'share_%': subtype_pct.round(2)
})

print(subtype_summary)

                        count  share_%
PropertySubType                       
SingleFamilyResidence  394249    72.93
Condominium             97835    18.10
Townhouse               31502     5.83
ManufacturedOnLand       7574     1.40
Duplex                   3376     0.62
StockCooperative         1716     0.32
NaN                      1256     0.23
Cabin                     953     0.18
Triplex                   671     0.12
MixedUse                  493     0.09
Quadruplex                263     0.05
MobileHome                163     0.03
BoatSlip                  120     0.02
OwnYourOwn                 93     0.02
ManufacturedHome           70     0.01
CoOwnership                64     0.01
Loft                       64     0.01
Farm                       59     0.01
Timeshare                  51     0.01
Studio                     33     0.01
DeededParking               6     0.00
Apartment                   1     0.00


> The dataset is **100% Residential** — the upstream filter was already applied.  
> Breakdown by `PropertySubType'

| Property Sub Type | Count | Share |
|-------------------|-------|-------|
| `SingleFamilyResidence` | 394,249 | ~73% |
| Condominium & others | ~146,363 | ~27% |

---

### Q2. What are the median and average close prices?

> `ClosePrice` is populated for **141,190 records only** (~26% of total — closed listings only).

| Metric | Value |
|--------|-------|
| **Median Close Price** | **$850,000** |
| **Mean Close Price** | **$1,200,000** |
| Gap | ~$350,000 (mean is ~41% above median) |

> The gap is driven by high-end luxury sales pulling the mean upward. For market reporting, **median is the preferred benchmark**.

### Q3. What does the Days on Market distribution look like?

In [28]:
#  Days on Market Distribution ──────────────────────────────────────────
dom = listed['DaysOnMarket']

print("=== Days on Market ===")
print(dom.describe().round(1))
print(f"\nNegative Days on Market (invalid):  {(dom < 0).sum()}")
print(f"Days on Market 0–30 days:           {((dom >= 0) & (dom <= 30)).sum()} ({((dom >= 0) & (dom <= 30)).mean()*100:.1f}%)")
print(f"Days on Market 31–90 days:          {((dom > 30) & (dom <= 90)).sum()} ({((dom > 30) & (dom <= 90)).mean()*100:.1f}%)")
print(f"Days on Market 91–180 days:         {((dom > 90) & (dom <= 180)).sum()} ({((dom > 90) & (dom <= 180)).mean()*100:.1f}%)")
print(f"Days on Market > 180 days:          {(dom > 180).sum()} ({(dom > 180).mean()*100:.1f}%)")

=== Days on Market ===
count    540612.0
mean         19.1
std          26.7
min         -58.0
25%           5.0
50%          11.0
75%          22.0
max         731.0
Name: DaysOnMarket, dtype: float64

Negative Days on Market (invalid):  29
Days on Market 0–30 days:           456226 (84.4%)
Days on Market 31–90 days:          66988 (12.4%)
Days on Market 91–180 days:         15431 (2.9%)
Days on Market > 180 days:          1938 (0.4%)


**Days on Market Distribution**
- 84% of listings sell within 30 days : market moves quickly
- Median: 11 days | Mean: 19.1 days : right-skewed, use median for reporting
- Negative DOM (29 records) and DOM = 731 days : remove during cleaning

### Q4. What percentage of homes sold above vs. below list price?

In [29]:
#Sold Above vs. Below List Price ───────────────────────────────────────
# ⚠️ Listing data only — ClosePrice exists for ~26% of records (closed listings only)
# This analysis is limited to that subset
sold = listed.dropna(subset=['ClosePrice', 'ListPrice']).copy()
sold['price_ratio'] = sold['ClosePrice'] / sold['ListPrice']

above = (sold['price_ratio'] > 1).sum()
at    = (sold['price_ratio'] == 1).sum()
below = (sold['price_ratio'] < 1).sum()
total = len(sold)

print("\n=== Sold Price vs. List Price (closed listings only, n={:,}) ===".format(total))
print(f"Above list price:  {above:,} ({above/total*100:.1f}%)")
print(f"At list price:     {at:,}  ({at/total*100:.1f}%)")
print(f"Below list price:  {below:,} ({below/total*100:.1f}%)")
print(f"\nMedian Close/List ratio: {sold['price_ratio'].median():.4f}")
print(f"Mean Close/List ratio:   {sold['price_ratio'].mean():.4f}")




=== Sold Price vs. List Price (closed listings only, n=141,190) ===
Above list price:  61,717 (43.7%)
At list price:     26,295  (18.6%)
Below list price:  53,178 (37.7%)

Median Close/List ratio: 1.0000
Mean Close/List ratio:   1.0763


### Q5. Are there any apparent date consistency issues (e.g., close date before listing date)?

In [30]:
# Date Consistency Issues ───────
listed['CloseDate_dt']    = pd.to_datetime(listed['CloseDate'], errors='coerce')
listed['ListingDate_dt']  = pd.to_datetime(listed['ListingContractDate'], errors='coerce')
listed['PurchaseDate_dt'] = pd.to_datetime(listed['PurchaseContractDate'], errors='coerce')

# Close before listing
c_before_l = listed.dropna(subset=['CloseDate_dt', 'ListingDate_dt'])
close_before_list = (c_before_l['CloseDate_dt'] < c_before_l['ListingDate_dt']).sum()

# Purchase before listing
p_before_l = listed.dropna(subset=['PurchaseDate_dt', 'ListingDate_dt'])
purchase_before_list = (p_before_l['PurchaseDate_dt'] < p_before_l['ListingDate_dt']).sum()

# Close before purchase
c_before_p = listed.dropna(subset=['CloseDate_dt', 'PurchaseDate_dt'])
close_before_purchase = (c_before_p['CloseDate_dt'] < c_before_p['PurchaseDate_dt']).sum()

print("\n=== Date Consistency Issues ===")
print(f"Close date before listing date:    {close_before_list}")
print(f"Purchase date before listing date: {purchase_before_list}")
print(f"Close date before purchase date:   {close_before_purchase}")




=== Date Consistency Issues ===
Close date before listing date:    72
Purchase date before listing date: 271
Close date before purchase date:   263


### Q6.Which counties have the highest median prices?

In [31]:
# Median Price by County
# ListPrice — all records
county_list = (
    listed.groupby('CountyOrParish')['ListPrice']
    .agg(median_list_price='median', total_listings='count')
    .sort_values('median_list_price', ascending=False)
    .reset_index()
)
print("\n=== Median List Price by County (Top 15, all listings) ===")
print(county_list.head(15).to_string(index=False))

# ClosePrice — closed listings only
# Limited to ~26% of records
county_close = (
    listed.dropna(subset=['ClosePrice'])
    .groupby('CountyOrParish')['ClosePrice']
    .agg(median_close_price='median', sold_count='count')
    .sort_values('median_close_price', ascending=False)
    .reset_index()
)
print("\n=== Median Close Price by County (Top 15, closed only) ===")
print(county_close.head(15).to_string(index=False))


=== Median List Price by County (Top 15, all listings) ===
 CountyOrParish  median_list_price  total_listings
      San Mateo          1648950.0            9933
    Santa Clara          1499000.0           26124
          Marin          1295000.0             229
     Santa Cruz          1250000.0            4553
         Orange          1200000.0           51723
  San Francisco          1098000.0            1533
        Alameda           998000.0           25861
       Monterey           994000.0            5225
    Los Angeles           950000.0          137333
      San Diego           910000.0           67640
San Luis Obispo           900000.0            6495
           Napa           899900.0             207
        Ventura           899000.0           15346
           Mono           850000.0              37
   Contra Costa           810000.0           24729

=== Median Close Price by County (Top 15, closed only) ===
 CountyOrParish  median_close_price  sold_count
      San Mateo 